Content.Server/GameTicking/Rules/RevolutionaryRuleSystem.cs

```diff
@@ -0,0 +1,463 @@
+        SubscribeLocalEvent<HeadRevolutionaryComponent, AfterFlashedEvent>(OnPostFlash);
+    }
+
+    protected override void Started(EntityUid uid, RevolutionaryRuleComponent component, GameRuleComponent gameRule, GameRuleStartedEvent args)
+    {
+        base.Started(uid, component, gameRule, args);
+        _endRoundCheck = _timing.CurTime + _timerWait;
+    }
+
+    /// <summary>
+    /// Checks if the round should end and also checks who has a mindshield.
+    /// </summary>
+    protected override void ActiveTick(EntityUid uid, RevolutionaryRuleComponent component, GameRuleComponent gameRule, float frameTime)
+    {
+        base.ActiveTick(uid, component, gameRule, frameTime);
+        if (GameTicker.IsGameRuleAdded(uid, gameRule) && _endRoundCheck <= _timing.CurTime)
+        {
+            _endRoundCheck = _timing.CurTime + _timerWait;
+            CheckFinish();
+        }
+    }
+
+    private void OnRoundEndText(RoundEndTextAppendEvent ev)
+    {
+        foreach (var headrev in EntityQuery<RevolutionaryRuleComponent>())
+        {
+            if (headrev.HeadsDied && !headrev.RevsLost)
+            {
+                ev.AddLine(Loc.GetString("rev-won"));
+            }
+            if (!headrev.HeadsDied && headrev.RevsLost)
+            {
+                ev.AddLine(Loc.GetString("rev-lost"));
+            }
+            if (headrev.HeadsDied && headrev.RevsLost)
+            {
+                ev.AddLine(Loc.GetString("rev-stalemate"));
+            }
+            if (!headrev.HeadsDied && !headrev.RevsLost)
+            {
+                ev.AddLine(Loc.GetString("rev-reversestalemate"));
+            }
+            ev.AddLine(Loc.GetString("head-rev-initial-count", ("initialCount", headrev.HeadRevs.Count)));
+            foreach (var player in headrev.HeadRevs)
+            {
+                ev.AddLine(Loc.GetString("head-rev-initial",
+                    ("name", player.Key),
+                    ("username", player.Value)));
+            }
+            break;
+        }
+    }
+    private void OnStartAttempt(RoundStartAttemptEvent ev)
+    {
+        var query = AllEntityQuery<RevolutionaryRuleComponent, GameRuleComponent>();
+        while (query.MoveNext(out var uid, out var revs, out var gameRule))
+        {
+            if (!GameTicker.IsGameRuleAdded(uid, gameRule))
+                continue;
+            if (!ev.Forced && ev.Players.Length < revs.MinPlayers)
+            {
+                _chatManager.SendAdminAnnouncement(Loc.GetString("rev-not-enough-ready-players",
+                    ("readyPlayersCount", ev.Players.Length),
+                    ("minimumPlayers", revs.MinPlayers)));
+                ev.Cancel();
+                continue;
+            }
+            if (ev.Players.Length == 0)
+            {
+                _chatManager.DispatchServerAnnouncement(Loc.GetString("rev-no-one-ready"));
+                ev.Cancel();
+            }
+        }
+    }
```

Comment:
Please don't copy code, pull it out into a different method and have the other gamerules use it instead.